### ISSUE TITLE
Add `@required` annotation for exported fields

### ISSUE THREAD

**Issue by kerelape** [godotengine/godot-proposals#8343]
### Describe the project you are working on

3D adventure

### Describe the problem or limitation you are having in your project

I am heavily using `@export var node: Node` where `Node` is any `Node` descendant to avoid depending much on node paths, but sometimes s.. happens and I simply forget to set a value to the field in the inspector, and in such cases I will only know about that in runtime.

### Describe the feature / enhancement and how it helps to overcome the problem or limitation

It would be great to have something like an annotation (e.g. `@required` or `@export_required`) 

### Describe how your proposal will work, with code, pseudo-code, mock-ups, and/or diagrams

When a property is marked as _required_ the inspector would promt you about it the same way configuration warnings do, but also highlight the missing property (e.g. the yellow triangle may be placed in front of property saying it's required to be set to a value).

This would not only work for nodes, but also for any `Object` descendant, like `Resource`.

### If this enhancement will not be used often, can it be worked around with a few lines of script?

A possible workaround would be to mark scripts with `@tool` and check for the properties in `_get_configuration_warnings` 

### Is there a reason why this should be core and not an add-on in the asset library?

It requires changes to gdscript

**Comment by Calinou** [comment]
I would prefer something like https://github.com/godotengine/godot-proposals/issues/1729, which is a more universal solution and doesn't error if the variable isn't used anywhere.

**Comment by kerelape** [comment]
> I would prefer something like #1729, which is a more universal solution and doesn't error if the variable isn't used anywhere.

This would not solve the issue with `Resource`-type properties.

And still the issue would only arrive when the game is already running.

What I want is to have a built-in problem reporting for exported properties the same way it's done for physics objects missing collision nodes

**Comment by RedMser** [comment]
> the inspector would promt you about it the same way configuration warnings do, but also highlight the missing property (e.g. the yellow triangle may be placed in front of property saying it's required to be set to a value).

You could use configuration warnings combined with https://github.com/godotengine/godot/pull/68420

**Comment by kerelape** [comment]
> > the inspector would promt you about it the same way configuration warnings do, but also highlight the missing property (e.g. the yellow triangle may be placed in front of property saying it's required to be set to a value).
> 
> You could use configuration warnings combined with [godotengine/godot#68420](https://github.com/godotengine/godot/pull/68420)

Yes, this is what I am currently doing, but it requires me make my scripts `@tool`, which requires me to add those `Engine.is_editor_hint()` checks. Also it's a quite frequent case and almost every script in my game has that line-to-line similar are-properties-set checks, which could be nice to simplify to a single annotation.

I think it's an overkill to do all those check by hand for a simple is-null validation. [68420](https://github.com/godotengine/godot/pull/68420) is great for heavy cases, where you do more than that

**Comment by miv391** [comment]
`@export` is the way that nodes need be to referenced. `@export` was most likely not invented for exactly this kind of use, but until Godot gets a feature that automatically fixes the following code when camera node's name changes or the node is moved, `@export` is so much better than getting nodes by name. (Yes, I know `%` is meant for fixing breaks caused by moving nodes, but it doesn't help when node is renamed.)

```gdscript
var cam1 = get("sub/camera")
var cam2 = $sub/camera
```

So it would be nice to have the proposed feature. I would guess that getting new features to export variables would be more possible than automatically renaming node paths.

Currently I'm doing something like this:

```gdscript
@export var camera: Camera2D

func _ready():
	assert(camera)
```

The feature could also be something like automatically asserting on `_ready` that the variable is not null. The sooner you get the error, the better.

**Comment by Pizzaandy** [comment]
Similar to #2519 

Right now, the intended way to customize inspector properties is to either:
a) Override `_get_property_list`
b) Write an inspector plugin

It's not very ergonomic to implement reusable property editors at the moment. 
I like the way Unity allows users to create property attributes and drawers, and I think that's the right way forward for Godot.

Give users the power to make custom annotations that change how exported properties are rendered in the inspector, then provide a few built-in validation annotations like `@required`.

This doesn't necessarily have to be implemented via annotations, but there should be *some* way of declaratively composing inspectors rather than messing with `_get_property_list` or plugins.

**Comment by nlupugla** [comment]
What about a slightly more general approach where you can tag properties as described in this proposal: https://github.com/godotengine/godot-proposals/issues/8752

Instead of writing

```gdscript
@required
@export var node : Node
```

you could write

```gdscript
@tool

@tag("required")
@export var node : Node :
    set(value : Node) :
        node = value
        update_configuration_warnings()


func _get_configuration_warnings() -> PackedStringArray:
    var warnings : PackedStringArray = []
    var properties : Array[Dictionary] = get_tagged_properties("required")
    for property in properties:
        if get(property.name) == null:
            warnings.append(property.name + " is required")
```

**Comment by leonardoraele** [comment]
> What about a slightly more general approach where you can tag properties as described in this proposal: #8752
> 
> [...]

Because then you have to reimplement this feature in every script you create.

**Comment by Nohac** [comment]
I would also like to see this in some form. but think this could be implicit whenever https://github.com/godotengine/godot-proposals/issues/162 is implemented.
If we get nullable types in godot, then any type that is not marked as nullable could show up as a required field and show a warning in the editor if it's not set.

I understand that nullable types would be a breaking change, so I hope they go the same rout as they did with static typing and make it an opt in setting.

**Comment by 0-Vanes-0** [comment]
> Currently I'm doing something like this:
> ```
> @export var camera: Camera2D
> 
> func _ready():
> 	assert(camera)
> ```
> The feature could also be something like automatically asserting on `_ready` that the variable is not null. The sooner you get the error, the better.

This thing is still much more preferable solution over `@onready` and `%`, yet having a big list of `@export` variables gives a big assert's parameter size:
```gdscript
# Example of big @export list
@export_group("Required Children")
@export var game_zone: Marker3D
@export var play_deck: PlayDeck
@export var garbage_deck: GarbageDeck
@export var cards_queue: CardsQueue
@export var player_storages_collection: PlayerStorageCollectionNode
@export var remote_cameras_collection: RemoteCameraCollectionNode
@export var spawner: PlayerSpawner

func _ready() -> void:
	assert(game_zone and play_deck and garbage_deck and cards_queue and player_storages_collection and remote_cameras_collection and spawner)
	# using this vars in the class or in other scripts freely...
	# without any null checks or node's name dependency
```
This functionality is the best in such cases, it's just a matter of comfort (duplicating variables in assert function **instead of** writing something like `@export_required` and getting error that stops a game and highlights a var in case of some export var is missing).
